components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java

```diff
@@ -4130,4 +4140,64 @@ protected void checkAccessControlPermission(APIIdentifier identifier) throws API
         return updatedWSDLContent;
     }
 
+    /**
+     * This method is used to get keys of custom attributes, configured by user
+     *
+     * @param userId user name of logged in user
+     * @return Array of JSONObject, contains keys of attributes
+     * @throws APIManagementException
+     */
+    public JSONArray getAppAttributesFromConfig(String userId) throws APIManagementException {
+
+        String tenantDomain = MultitenantUtils.getTenantDomain(userId);
+        int tenantId = 0;
+        try {
+            tenantId = getTenantId(tenantDomain);
+        } catch (UserStoreException e) {
+            handleException("Error in getting tenantId of " + tenantDomain, e);
+        }
+        JSONArray applicationAttributes = null;
+        JSONObject applicationConfig = APIUtil.getAppAttributeKeysFromRegistry(tenantId);
+        try {
+            if (applicationConfig != null) {
+                applicationAttributes = (JSONArray) applicationConfig.get(APIConstants.ApplicationAttributes.ATTRIBUTES);
+            } else {
+                APIManagerConfiguration configuration = getAPIManagerConfiguration();
+                applicationAttributes = configuration.getApplicationAttributes();
+            }
+        } catch (NullPointerException e){
+            handleException("Error in reading configuration " + e.getMessage(), e);
+        }
+        return applicationAttributes;
+    }
+
+    /**
+     * This method is used to validate keys of custom attributes, configured by user
+     *
+     * @param application
+     * @param userId user name of logged in user
+     * @throws APIManagementException
+     */
+    public void checkAppAttributes(Application application, String userId) throws APIManagementException {
+
+        JSONArray applicationAttributesFromConfig = getAppAttributesFromConfig(userId);
+        Map<String, String> applicationAttributes = application.getApplicationAttributes();
+        List attributeKeys = new ArrayList<String>();
+        int applicationId = application.getId();
+
+        for (Object object : applicationAttributesFromConfig) {
+            JSONObject attribute = (JSONObject) object;
+            attributeKeys.add(attribute.get(APIConstants.ApplicationAttributes.ATTRIBUTE));
+        }
+
+        for (Object key : applicationAttributes.keySet()) {
```

Comment:
This one too?

Reply:
If there are no attributes defined in configuration, the size of JSONArray would be zero. Then this for-loop won't get executed. Also, this array is being initialized here [1]. Hence, it will not produce NPE.

[1] https://github.com/wso2/carbon-apimgt/blob/6.x/components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java#L82